Portable PDB support for ilasm

ivanpovazan opened the pull request:
This is my take on adding portable PDB support to ilasm.

I understand that this PR is quite large, which is against the contribution guidelines, but adding portable PDB support requires a lot of changes.

Even though the work is not complete, after talking to @noahfalk I have decided to create this draft pull request to start the discussion and get some feedback on the work that has been done so far. Thus I would really appreciate any comments/suggestions/corrections in order to finalize this feature.

The work is based on the thread discussion (#5051) and mainly follows the idea of creating a portable PDB metadata writer based on extensions to IMetaDataEmit interface. It is still WIP, but the current version can generate portable PDB, which can be loaded by VS debugger and I have successfully managed to debug, step, observe local variables of C# programs by round tripping them: compiling -> disassembling (ildasm) -> assembling(ilasm).

Since portable PDB standard seems to be oriented more towards higher level programming languages, I have tried to define minimum viable set of features that are applicable to ilasm, which are currently supported:
- Portable PDB CodeView debug directory entry
- Document table
- MethodDebugInformation table
- Local scope table
- Local variable table

Please let me know if some other features are necessary or required, and how would those be expressed in ilasm.

Commit:
[WIP] Adding new command line argument to specify desired PDB format

- Currently this just makes a distinction between classic PDB format (not supported on CoreCLR) and portable one (to be added)

Commit:
[WIP] Extending IMetaDataDispenserEx to create a metadata emitter used for portable PDB generation

- The idea is to have 2 emitters in ilasm, 1 for system metadata and 1 for pdb metadata

Commit:
[WIP] Adding PortablePdbWritter class as a wrapper around emitter used for pdb metadata generation

Commit:
[WIP] Adding methods for setting PDB file name and saving PDB file

Commit:
[WIP] Updating CreateDebugDirectory method to support portable PDB

Commit:
[WIP] Adding portable PDB table definitions

StateMachineMethod and CustomDebugInformation tables are left out

Commit:
[WIP] Adding PdbStream generation

This change extends the IMetaDataEmit2 interface

Commit:
[WIP] Adding the logic for defining documents

Extending IMetaDataEmit2 interface

Commit:
[WIP] Adding the logic for defining sequence points

Extending IMetaDataEmit2 interface

Commit:
[WIP] Make sure String heap is never empty for portable PDB metadata generation

- This can probably be done in a better (more clever) way, check if/why/when this is necessary

Commit:
[WIP] Adding the logic for defining local scopes and local variables

Extending IMetaDataEmit2 interface

Commit:
[WIP] Removing line continuation in comments

Fixes build failures under gcc -Wcomment -Werror

davidwrighton reviewed (changes requested):
I like the ideas here, and the implementation generally seems reasonable. I have a few concerns though, that I'd like to make sure are handled before we could accept such a change. At

1. All of the changes in the MD directory need to be under some #ifdef. We could enable that for building ilasm, and probably ildasm, as ildasm should probably be able to parse these things. However, the size of the coreclr binary must not be affected by these changes.
2. I'd like to avoid changes to in the public headers Cor.h and Corhdr.h. I'd prefer that all new portable pdb handling in the native api is held in a new header or two.
3. COM rules on interface changing must be respected. This is a minor detail, and easy to fix.
4. We should support both embedded and non-embedded portable pdbs.
5. This will need some amount of testing.

noahfalk commented:
Hi @ivanpovazan - thanks so much for your efforts here! It might be another day or two before I can dig into this properly.
heads up @tmat and @MichalStrehovsky, you guys might be interested in taking a look at this as well?

Commit:
[WIP] Linux build failure fix

ivanpovazan commented on src/coreclr/src/inc/cor.h:
```diff
@@ -613,6 +631,41 @@ DECLARE_INTERFACE_(IMetaDataEmit2, IMetaDataEmit)
 
     STDMETHOD(ResetENCLog)() PURE;          // S_OK or error.
 
+    STDMETHOD(GetReferencedTypeSysTables)(  // S_OK or error.
```
Thank you for your insightful feedback.
I am currently fixing some build issues, but will look into introducing new COM interfaces right after that.

ivanpovazan replied:
@davidwrighton could you please elaborate on how would these new COM interfaces be generated?
I was thinking of creating ppdb.idl file which would define IMetaDataEmit3 interface which would inherit from IMetaDataEmit2, but I failed to find an .idl file defining IMetaDataEmit2

davidwrighton replied:
These COM interface headers were written by hand. I would recommend following the pattern in cor.h to make another one. 

For instance, 
```
// {7DD2F810-CAD1-419F-9964-82912CAC3CD4}
EXTERN_GUID(IID_IMetaDataEmit3, 0x7DD2F810, 0xCAD1, 0x419F, 0x99, 0x64, 0x82, 0x91, 0x2C, 0xAC, 0x3C, 0xD4);
//---
#undef  INTERFACE
#define INTERFACE IMetaDataEmit3
DECLARE_INTERFACE_(IMetaDataEmit3, IMetaDataEmit2)
{
// Insert new api's here!
};
```

Note, that I have generated a new guid by using the uuidgen tool. Each guid is associated represents a unique interface.

ivanpovazan replied:
@davidwrighton could you please check the changes from my last 3 commits?

Based on your previous comments, I have done the following:
1. Introduced 2 new COM interfaces: _IMetaDataEmit3_ and _IMetaDataDispenserEx2_ found in .\md\inc\portablepdbmdi.h
2. Removed all changes from cor.h and corhdr.h
3. Introduced new preprocessor directive _FEATURE_METADATA_EMIT_PORT_PDB_ which now surrounds all the changes related to portable PDB metadata generation
4. Adjusted make files to generate portable PDB metadata support only for ilasm (keeping coreclr intact)

Let me know what do you think about this, and what will be the next steps.

Commit:
[WIP] Another attempt to fix build issues for different systems

Commit:
[WIP] Proper use of IfFailGo macro for checked build

janvorli commented on src/coreclr/src/md/enc/pdbheap.cpp:
```diff
@@ -30,13 +30,13 @@ HRESULT PdbHeap::SetData(PORTABLE_PDB_STREAM* data)
 
     ULONG offset = 0;
-    IfFailGo(hr = memcpy_s(m_data + offset, m_size, &data->id, sizeof(data->id)));
+    IfFailGo(memcpy_s(m_data + offset, m_size, &data->id, sizeof(data->id)));
```
memcpy_s doesn't return `HRESULT`, it returns `errno_t`. So you cannot use IfFailGo for it.

ivanpovazan replied:
Thank you for your remark, this has been fixed with https://github.com/dotnet/runtime/pull/37702/commits/b99bdf5685d18eea23343fe3e638d7f5e5cf41f8

noahfalk commented:
@janvorli and @davidwrighton - I've had plenty of work continue to come in and I see both of you already reviewed. Let me know if you think it would valuable for me to be a 3rd reviewer here.

Commit:
[WIP] Removing inappropriate use of IfFailGo macro

Commit:
[WIP] Introducing new COM interfaces for portable PDB metadata generation

All changes related to portable PDB metadata generation should be controlled with preprocessor directive FEATURE_METADATA_EMIT_PORT_PDB

Commit:
[WIP] Surrounding all portable PDB metadata changes with FEATURE_METADATA_EMIT_PORT_PDB preprocessor directive

Commit:
[WIP] Enabling only ilasm to use portable PDB metadata generation

Commit:
[WIP] Fix for UNIX paths when defining documents

davidwrighton commented:
@ivanpovazan I've looked through the change, and I believe it is good enough to checkin, and there are no more meaningful issues to address before checkin with regards to changes to the product code. I am however a bit concerned that it might regress some existing ILasm functionality as we have very few tests for ilasm correctness at this time.

@BruceForstall do you have any ideas on how to regresssion test this? I know we've done some ildasm/ilasm roundtripping testing in the past. Is that a reasonable thing to run?

BruceForstall commented on src/coreclr/src/ilasm/portable_pdb.cpp:
```diff
@@ -0,0 +1,396 @@
+    memset(m_pdbStream.typeSystemTableRows, 0, sizeof(ULONG) * TBL_COUNT);
+    time_t now;
+    time(&now);
+    m_pdbStream.id.pdbTimeStamp = (ULONG)now;
```
Do we really want the timestamp in the generated PDB file? That makes the generated file not binary equivalent on equivalent source rebuild. I thought we're trying to get away from timestamps in generated files because of this problem.

BruceForstall commented on src/coreclr/src/ilasm/portable_pdb.cpp:
```diff
@@ -0,0 +1,396 @@
+    time_t now;
+    time(&now);
+    m_pdbStream.id.pdbTimeStamp = (ULONG)now;
+    hr = CoCreateGuid(&m_pdbStream.id.pdbGuid);
```
Maybe the same this applies to this GUID.

BruceForstall commented:
> @BruceForstall do you have any ideas on how to regresssion test this? I know we've done some ildasm/ilasm roundtripping testing in the past. Is that a reasonable thing to run?

Adding back ilasm/ildasm round trip testing is still a work item (which I recently assigned to myself): https://github.com/dotnet/runtime/issues/36599. 

Even if done, it would only test that basic functionality was ok, without portable PDB usage (still an important scenario to test).

Round trip testing with `/DEBUG /PDBFMT=PORT` would presumably test the new functionality. Although ildasm wouldn't look at the PDB, so the round tripping presumably wouldn't test much. (Does ildasm need 

Simply replacing ilasm in the build of the coreclr tests and forcing `/DEBUG /PDBFMT=PORT` would at least test the PDB generation.

So how to test that the generated PDB is correct and debuggable? Do we also need to test `/DEBUG=IMPL` and `/DEBUG=OPT`?

What corresponding changes should be made to ildasm? When ildasm was open sourced and ported from Windows and made cross-platform, PDB support was dropped.  So, for instance, `g_pSymReader` is never initialized (even though lots of code that could use it still exists). Should this be brought back?

BruceForstall commented:
> @janvorli and @davidwrighton - I've had plenty of work continue to come in and I see both of you already reviewed. Let me know if you think it would valuable for me to be a 3rd reviewer here.

@noahfalk There are big enough diagnostics-related changes and issues here that I think it would be valuable for you to look.

davidwrighton commented:
@BruceForstall I don't believe its reasonable add a dependency of making portable pdbs work with ildasm required to get this work checked in, although it would actually be really nice to have. Perhaps we could have a test in our test suite which uses the new ilasm, produces a pdb, runs it, throws and exception, and shows that the runtime is able to read the pdb correctly?

BruceForstall commented:
>  I don't believe its reasonable add a dependency of making portable pdbs work with ildasm required to get this work checked in

Agreed. I assume the extra output ildasm would produce if reading a PDB would not contribute to proper round-tripping?

> Perhaps we could have a test in our test suite which uses the new ilasm, produces a pdb, runs it, throws and exception, and shows that the runtime is able to read the pdb correctly?

An interesting idea. As long as it can be written to be robust!

ivanpovazan commented:
@davidwrighton @BruceForstall thank you for your remarks. 
I am listing the responses to your comments:

- **naming**: I will fix the naming issues as suggested.
- **PDB timestamp and GUID**: I was just following the specification of [debug directory](https://github.com/dotnet/runtime/blob/master/docs/design/specs/PE-COFF.md) and [portable PDB format](https://github.com/dotnet/runtime/blob/master/docs/design/specs/PortablePdb-Metadata.md).
- **testing**: I am currently writing some tests which use _System.Reflection.Metadata_ library to verify the portable PDB metadata format and will commit my work as soon as possible.
- **more on testing**: For the other testing topics like ilasm regression, loading/debugging of the portable PDB metadata with a debugger; I am afraid that currently I don't have enough information on how to proceed with that

Let me know what do you think about this.

davidwrighton commented:
With regards to actually using the debugging format in the runtime/debugger, if instead you've written reasonable quality System.Reflection.Metadata tests for the portable format that can be part of our CI system, and we've done some manual validation that the pdb is useable in a debugger/runtime context, I don't think that we need to write the test I proposed. The important part is that whatever tests you write need to somehow run as part of our regular testing, which may be difficult as ilasm is currently an undertested portion of our product. However, if you can get tests written, I will find a way to run them in our tests.

BruceForstall commented:
btw, I implemented ildasm/ilasm round-trip testing support as a new Azure DevOps Pipeline (https://github.com/dotnet/runtime/pull/38490), so at least we'll have the ability to do some testing once that's merged.

Commit:
[WIP] Fix for local scope generation

Commit:
[WIP] Adding functional tests for ilasm portable PDB support

ivanpovazan commented:
@davidwrighton @BruceForstall my last commit includes some functional tests related to portable PDB metadata generation for ilasm.

It is true that testing would be even more complete when ildasm supports portable PDB, but I believe I covered significant part of what has been added for portable PDB support.

What I think would be valuable for you (or someone who will perform manual debugger tests) can be found at:
https://github.com/ivanpovazan/runtime/tree/ilasm_portable_pdb/src/coreclr/tests/src/ilasm/PortablePdb/Resources
where I have added a short description of how you can use one of the test files to verify debugging C# source with a debugger.

Commit:
[WIP] Spelling out PORTABLE correctly in preprocessor directive controlling portable PDB metadata support

Commit:
[WIP] Typo: Naming PortablePdbWriter instead of PortablePdbWritter

Commit:
[WIP] Clarifying command line parameters for pdb format by using their full names

ivanpovazan commented on src/coreclr/src/ilasm/CMakeLists.txt:
```diff
@@ -4,6 +4,7 @@ add_definitions(-DUNICODE)
 add_definitions(-D_UNICODE)
 add_definitions(-D_FEATURE_NO_HOST)
 add_definitions(-D__ILASM__)
+add_definitions(-DFEATURE_METADATA_EMIT_PORT_PDB)
```
https://github.com/dotnet/runtime/pull/37702/commits/a8c264928fa4c78c3361cf006d9881279797d8fb

ivanpovazan commented on src/coreclr/src/ilasm/assembler.h:
```diff
@@ -1050,6 +1060,20 @@ class Assembler {
+    // Portable PDB paraphernalia
+public:
+    PdbFormat           m_pdbFormat;
+    PortablePdbWritter* m_pPortablePdbWritter;
```
Fixed with: https://github.com/dotnet/runtime/pull/37702/commits/9c9b3f5b6175c8c772dc0dd5d66e87fd9efd9b68

ivanpovazan commented on src/coreclr/src/ilasm/main.cpp:
```diff
@@ -164,6 +167,8 @@ extern "C" int _cdecl wmain(int argc, __in WCHAR **argv)
       printf("\n/EXE            Compile to .exe (default)");
       printf("\n/PDB            Create the PDB file without enabling debug info tracking");
+      printf("\n/PDBFMT=CLA     Use classic PDB format for PDB file generation (default)");
+      printf("\n/PDBFMT=PORT    Use portable PDB format for PDB file generation");
```
Fixed with: https://github.com/dotnet/runtime/pull/37702/commits/a1641560f7c1f2de6ac89d2599dc838dc22686dc

ivanpovazan commented:
I have also fixed the naming issues reported by @BruceForstall.
Can I mark this as 'Ready for review' so this work can be treated as done?

davidwrighton commented:
Please feel free to mark this as ready for review. I think the testing strategy here is good. I'm not sure the tests are in the right place, but I don't have a better place for them quite yet.

ivanpovazan commented:
@davidwrighton thank you, I have marked it as ready for review but there is still a change request open.

noahfalk commented:
> * **PDB timestamp and GUID**: I was just following the specification of [debug directory](https://github.com/dotnet/runtime/blob/master/docs/design/specs/PE-COFF.md) and [portable PDB format](https://github.com/dotnet/runtime/blob/master/docs/design/specs/PortablePdb-Metadata.md).
> 
> Let me know what do you think about this.

This looks fine to me. I didn't notice anything in the portable pdb spec that said the 20 byte id needed to be formed from a GUID and a timestamp, but that behavior matches [SRM](https://github.com/dotnet/runtime/blob/0ac5dc9f7a8524dcec83de8f02a738d51b26f5f3/src/libraries/System.Reflection.Metadata/src/System/Reflection/Metadata/BlobContentId.cs#L87)'s implementation when it is operating in non-deterministic mode.

ivanpovazan commented:
@noahfalk yes you are right, GUID and timestamp requirement do not come from portable PDB spec, that is why I linked the spec of debug directory as well, which states:

> Version Major=any, Minor=0x504d of the data format has the same structure as above. The Age shall be 1. The format of the .pdb file that this PE/COFF file was built with is Portable PDB. The Major version specified in the entry indicates the version of the Portable PDB format. Together 16B of the Guid concatenated with 4B of the TimeDateStamp field of the entry form a PDB ID that should be used to match the PE/COFF image with the associated PDB (instead of Guid and Age). Matching PDB ID is stored in the #Pdb stream of the .pdb file.

noahfalk commented:
> that is why I linked the spec of debug directory as well...

I have failed at reading it appears, sorry about that : )

noahfalk reviewed (commented):
This looks really good to me, thanks for your all your hard work here! Just one comment inline about the verification macros I'd like to fix up.

ivanpovazan commented on src/coreclr/src/md/inc/VerifyLayouts.inc:
```diff
@@ -166,6 +166,9 @@ FIELD(CLiteWeightStgdbRW, m_wszFileName, sizeof(void*))
 FIELD(CLiteWeightStgdbRW, m_dwDatabaseLFS, 4)
 FIELD(CLiteWeightStgdbRW, m_pStgIO, sizeof(void*))
+#ifdef FEATURE_METADATA_EMIT_PORTABLE_PDB
+FIELD(CLiteWeightStgdbRW, m_pPdbHeap, sizeof(void*))
```
Correct, I was getting static assert failures.
I will check the make files to adjust this and will get back to you.

ivanpovazan replied:
Please take a look at my last commits which reflect this.

noahfalk replied:
looks great!

Commit:
[WIP] Removing dependency on layout verification from all mdcompiler library variations except for mdcompiler_wks

ivanpovazan committed:
Merge branch 'master' into ilasm_portable_pdb

Commit:
[WIP] Simplifying the makefile for building mdcompiler_wks library

Commit:
[WIP] Moving ilasm portable PDB tests to a new location to make them part of test suite

noahfalk reviewed (approved):
LGTM!

ivanpovazan commented:
Any idea how I could track down the unsuccessful checks? 

> .packages\microsoft.dotnet.helix.sdk\5.0.0-beta.20330.3\tools\Microsoft.DotNet.Helix.Sdk.MultiQueue.targets#L76
> .packages\microsoft.dotnet.helix.sdk\5.0.0-beta.20330.3\tools\Microsoft.DotNet.Helix.Sdk.MultiQueue.targets(76,5): error : (NETCORE_ENGINEERING_TELEMETRY=Test) Work item e58d5f52-74bf-4657-85be-7893ea3587d9/System.Collections.Immutable in job e58d5f52-74bf-4657-85be-7893ea3587d9 has failed.
> Failure log: https://helix.dot.net/api/2019-06-17/jobs/e58d5f52-74bf-4657-85be-7893ea3587d9/workitems/System.Collections.Immutable/console

BruceForstall commented:
@ivanpovazan It appears your change hit the CI in a period of particularly poor reliability. My suggestion is to rebase and repush your change to trigger a new set of runs. (I've not had good luck simply re-triggering from the UI, or close/re-open to attempt to retrigger in cases like this.)

ivanpovazan commented:
Thank you for the explanation, I will try to retrigger the build

Commit:
Merge branch 'master' into ilasm_portable_pdb

Commit:
Merge branch 'master' into ilasm_portable_pdb

davidwrighton commented:
I've manually verified that debugging works, and this scheme for testing looks to be good too. Thank you @ivanpovazan for all your hard work.
